tensorflow-keras/src/main/java/org/tensorflow/keras/backend/K.java

```diff
@@ -0,0 +1,713 @@
+  private static <T extends TType, U extends TNumber> Operand moveDimToEnd(
+      Ops tf, Operand tensor, int dim_index, Operand rank) {
+    Operand one = one(tf, TInt32.DTYPE);
+    List<Operand<T>> concatList =
+        Arrays.asList(
+            tf.range(tf.constant(dim_index), one, one),
+            tf.range(tf.constant(dim_index + 1), rank, one));
+    return tf.linalg.transpose(
+        tensor,
+        (Operand<U>) tf.concat((Iterable<Operand<T>>) concatList, (Operand<U>) tf.constant(0)));
+  }
+
+  private static <T extends TType, U extends TNumber> Operand flattenOuterDims(
+      Ops tf, Operand logits) {
+    Operand zero = zero(tf, TInt64.DTYPE);
+    Operand one = one(tf, TInt64.DTYPE);
+    Operand minusOne = tf.constant(-1);
+
+    // Shape logitsShape = logits.asOutput().shape();
+    // long lastDimSize = logitsShape.size(logitsShape.numDimensions()-1);
+    // if(!tf.scope().env().isEager()) {
+    Shape shape = logits.asOutput().shape();
+    int ndims = shape.numDimensions();
+    if (!shape.hasUnknownDimension()) {
+      long product = 1L;
+      boolean productValid = true;
+      for (int i = ndims - 2; i >= 0; i--) {
+        long d = shape.size(i);
+        if (d == Shape.UNKNOWN_SIZE) {
+          productValid = false;
+          break;
+        }
+        product *= d;
+      }
+      if (productValid) {
+        Shape outputShape = Shape.of(product, shape.size(ndims - 1));
+        return tf.reshape(logits, tf.constant(outputShape.asArray()));
+      }
+    }
+    // }
+
+    Operand rank = tf.dtypes.cast(tf.rank(logits), TInt64.DTYPE);
+    Operand rankMinusOne = tf.math.sub(rank, one);
+
+    Operand last_dim_size = tf.slice(tf.shape(logits), rankMinusOne, tf.constant(1));
+    Operand concat =
+        tf.concat(Arrays.asList(tf.constant(new int[] {-1}), last_dim_size), tf.constant(0));
+    return tf.reshape(zero, concat);
+  }
+
+  private static int[] moveAxisToEnd(int axis, int outputRank) {
+    int[] axisNew = new int[outputRank];
+    for (int i = 0; i < axis; i++) {
+      axisNew[i] = i;
+    }
+    for (int i = axis + 1; i < outputRank; i++) {
+      axisNew[i - 1] = i;
+    }
+    axisNew[outputRank - 1] = axis;
+    return axisNew;
+  }
+
+  // TODO, maybe part of Shape ??
+  private static boolean shapeIsCompatible(Shape a, Shape b) {
+    if (a.numDimensions() != b.numDimensions()) {
+      return false;
+    }
+    for (int i = 0; i < a.numDimensions(); i++) {
+      long aSize = a.size(i);
+      long bSize = b.size(i);
+      if (aSize != Shape.UNKNOWN_SIZE && bSize != Shape.UNKNOWN_SIZE && aSize != bSize) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  // TODO these are "nn" ops
+  public static Operand softmax_cross_entropy_with_logits(Ops tf, Operand labels, Operand logits) {
+    return softmax_cross_entropy_with_logits(tf, labels, logits, -1);
+  }
+
+  public static Operand softmax_cross_entropy_with_logits(
```

Comment:
Python TensorFlow defines` tf.nn.softmax_cross_entropy_with_logits` in `nn_ops.py`. which calls `softmax_cross_entropy_with_logits_v2_helper` in the same file, (reimplemented in Java with this method),
`softmax_cross_entropy_with_logits_v2_helper` manipulates the shapes before calling `gen_nn_ops.softmax_cross_entropy_with_logits` (equivalent to the current `tf.nn.softmax_cross_entropy_with_logits` in Java). It appears that the low level op in python is not exposed to the high-level API but is fronted with python logic in `softmax_cross_entropy_with_logits_v2_helper`. 

I have found this pattern a couple of times in python where python code fronts the call to the C api and the actual C api is not exposed in the higher level TF api.

Reply:
Ok, so sounds like we should move the current one we've got into a less obvious place, rename it and replace the one in `tf.nn` with yours.

Reply:
I could overload `softmaxCrossEntropyWithLogits` by adding the axis parameter.
That would create 2 versions of `softmaxCrossEntropyWithLogits` though.

```
public static <U extends TType, T extends TNumber> Operand<T> `softmaxCrossEntropyWithLogits`(
            Scope scope, Operand<T> labels, Operand<U> logits, int axis)
```

Reply:
But then I cannot do this with `sparseSoftmaxCrossEntropyWithLogits`. We would need either to come up with a naming convention that distinguishes the two, or hide the low level version from NnOps. They would only be accessible through the higher level api, at least from NnOps.

Reply:
We can make `tf.nn.internal` and put the low level ones there. Or rename them to include internal in the name. @karllessard what do you think?

Reply:
I favor `tf.nn.internal`, so far this applies to  `softmaxCrossEntropyWithLogits` and `sparseSoftmaxCrossEntropyWithLogits`. There are probably more.

Reply:
I have made these changes adding `tf.nn.internal.softmaxCrossEntropyWithLogits()` and `tf.nn.internal. sparseSoftmaxCrossEntropyWithLogits` and created  `NN.java` in package `org.tensorflow.op.core` that creates
`tf.nn.softmaxCrossEntropyWithLogits()`  and `tf.nn.sparseSoftmaxCrossEntropyWithLogits() `for the higher level functionality. Both of these classes use the `org.tensorflow.op.nn.internal.SoftmaxCrossEntropyWithLogits` and `org.tensorflow.op.nn.internal.SparseSoftmaxCrossEntropyWithLogits` classes.

If this is OK with everyone, I can push these changes.

Reply:
I'm ok with moving raw ops that we front in a `.internal` subpackage (or maybe `tf.nn.raw`?) To do this @JimClarke5 , just have just updated the API def for that op, right?

The original initial idea though was to leave all ops in `Ops` as closest as possible to their raw implementation in the native library and to wrap them with high-level manipulations only in `tensorflow-framework`. I think what prevent us to do this efficiently is that the annotation processor right now only runs on top of `tensorflow-core-api`... I think we need to discuss more about this.

Reply:
Yes, I just updated the` api_def_SoftmaxCrossEntropyWithLogits.pbtxt` and `api_def_SparseSoftmaxCrossEntropyWithLogits.pbtxt`. Just let me know the agreed upon name of the new subpackage.

```
op {
  graph_op_name: "SoftmaxCrossEntropyWithLogits"
  endpoint {
    name: "nn.internal.SoftmaxCrossEntropyWithLogits"
  }
}
```

Reply:
I suggest we name it `raw` instead of `internal`, which reflects the same package name in Python.

Still, I think we need to discuss a bit more about it during our next community call.

Reply:
I just found another instance of a high level Python op on top of the raw op with the same name.
Python tensorflow, `tf.linalg.matmul() ` in` math_ops.py`, has python logic on top of the low level matmul Op in `gen_math_ops.py`.
All I know for now, is if I run python `numpy.tensordot()` or python `tf.tensordot()` on one set of inputs, I do not get the same results as when calling my java version of ` tensordot()` that merely ends up calling the current raw op `tf.lingalg.matMul()` in Java. I have isolated the difference to the missing higher level logic of python` tf.linalg.matmull()`in` math_ops.py.`

My purpose here is to bring attention to the pattern of higher level ops on top of the raw ops, so the issue I am seeing with `nn` is not a one off.

Reply:
Ok changed to `raw`

Reply:
I changed the package name to raw, and will be checking in the new NNops.java.